### ISSUE TITLE
OOM can cause linked array qs to break

### ISSUE THREAD

**Issue by franz1981** [JCTools/JCTools#355]
Not high priority given that I won't suggest users to run *any* application after receiving an OOM, but still relevant: [here](
https://github.com/netty/netty/issues/11956#issuecomment-1002041097) an explanation on how it could happen.

It could be *easily* solved by changing semantic of poll, peek and isEmpty during a resizing:
On an empty q, If the parity bit is just set and the new linked chunk is not yet set, isEmpty should return true and both poll and peek should return null.
This allow a producer index rollback to pass unnoticed because resizing won't make a consumer to believe that there's some data.
Data will be perceived as available just when producer index is further moved (and by consequence, a JUMP ready to be spin-awaited)

TLDR any emptiness check should use both consumer and producer index >> 1 to save considering any parity bit due to resizing

@nitsanw other ideas?

**Comment by nitsanw** [comment]
I think we can fix the spin loop to read the updated `pIndex` in the loop so it can recover, or we can take the current `size` policy and extend it. 

_\<rambling\> I'm a little uncomfortable with the recovery logic violating the monotonically increasing property of the indexes. It's a nice property to have, but I can also see the argument that the parity bit is not REALLY an advance in this usecase, we could have used the sign bit for the same purpose and that would have violated monotonic progress just as much.\</rambling\>_

Proposed solution, we follow the `size` current behaviour here:
```
size = (cIndex - pIndex)/2 -> queue in transition is empty

=> isEmpty = (cIndex - pIndex)/2 == 0 (make it match size, but without all the funky mitigations)

=> poll/peek loop needs to change:
            long pIndex = lvProducerIndex();
            if (index - pIndex)/2 == 0) return null;
            // poll() == null iff queue is empty, null element is not strong enough indicator, so we must
            // check the producer index. If the queue is indeed not empty we spin until element is
            // visible.
            do
            {
                e = lvRefElement(buffer, offset);
            }
            while (e == null);
```

**Comment by franz1981** [comment]
Agree: I like the proposed solution and I was thinking about re-reading pIndex every X spin for element as well (very happy we reached the same exact conclusions), although I believe that adding a chunk could be a slow operation (if the allocation is outside a tlab, for example) and being able to return back to the user application code (with a null poll) can help similarly to relaxedPoll to give the application code the chance to do something different (if any)
